### PR TITLE
Close #192: improving testing of management module, fixed some issues

### LINKED ISSUE #192: [mnm] More passthrough testing
- [x] create other test entities with registry and stats
- [x] test stats and contexts
- [x] test notifs and contexts

### PR DESCRIPTION
**Pull request by mathieucarbou**
@anthonydahanne : since this PR is big (but splitted into several atomic ones), I opened it up before it is finished so that you can start looking at it.

The only thing remaining is to complete the tests methods that currently fail on purpose.

This PR is to have a solid testing module and solid tests in tc-platform, based on passthrough first.

There's another issue logged to add integration tests by using the new sample entity inside this PR.

- I fixed 2 issues (see :bug: commits)
- I had to refactored a little some generic methods and the list containing the capabilities and descriptors so that we can have reproducible tests more easily.

### CODE REVIEW

**Comment by mathieucarbou** [comment]
test this please

File: management/testing/pom.xml [hunk 1]
```diff
@@ -27,13 +27,15 @@
   </parent>
 
   <modelVersion>4.0.0</modelVersion>
-  <groupId>org.terracotta.management.samples</groupId>
-  <artifactId>samples</artifactId>
-  <name>Terracotta Management :: Samples</name>
+  <groupId>org.terracotta.management.testing</groupId>
+  <artifactId>testing</artifactId>
```
Comment (anthonydahanne) [inline management/testing/pom.xml]: testing-parent instead for artifactId ? 
org.terracotta.management.testing:testing sounds weird and confusing to me

File: management/distribution/management-server/pom.xml [hunk 2]
```diff
@@ -89,8 +89,8 @@
                   <exclude>org.terracotta.management:sequence-generator</exclude>
                   <exclude>org.terracotta.management:cluster-topology</exclude>
                   <exclude>org.terracotta.management:management-registry</exclude>
-                  <exclude>org.terracotta.management.dist:management-common</exclude>
                   <exclude>org.terracotta:statistics</exclude>
+                  <exclude>org.terracotta.management.dist:management-common</exclude>
                   <exclude>org.terracotta:offheap-resource</exclude>
```
Comment (anthonydahanne) [inline management/distribution/management-server/pom.xml]: unnecessary diff ? (move 1 xml line)

File: management/management-registry/src/main/java/org/terracotta/management/registry/collect/StatisticsRegistryMetadata.java [hunk 3]
```diff
@@ -85,24 +85,32 @@ public StatisticsRegistryMetadata(StatisticsRegistry statisticsRegistry) {
...
         List<? extends Timestamped<? extends Number>> history = statistic.history(since);
         switch (statistic.type()) {
-          case COUNTER: return new CounterHistory(buildSamples(history, Long.class), NumberUnit.COUNT);
-          case RATE: return new RateHistory(buildSamples(history, Double.class), TimeUnit.SECONDS);
-          case LATENCY_MIN: return new DurationHistory(buildSamples(history, Long.class), TimeUnit.NANOSECONDS);
-          case LATENCY_MAX: return new DurationHistory(buildSamples(history, Long.class), TimeUnit.NANOSECONDS);
-          case LATENCY_AVG: return new AverageHistory(buildSamples(history, Double.class), TimeUnit.NANOSECONDS);
-          case RATIO: return new RatioHistory(buildSamples(history, Double.class), NumberUnit.RATIO);
-          case SIZE: return new SizeHistory(buildSamples(history, Long.class), MemoryUnit.B);
+          case COUNTER: return new CounterHistory(buildNaturalSamples(history), NumberUnit.COUNT);
```
Comment (anthonydahanne) [inline management/management-registry/src/main/java/org/terracotta/management/registry/collect/StatisticsRegistryMetadata.java]: test case to show the support of int stats for size for example ?

File: management/testing/entity-tests/src/test/java/org/terracotta/management/entity/sample/HelloWorldTest.java [hunk 4]
```diff
@@ -78,12 +59,13 @@ public void tearDown() throws Exception {
   }
 
   @Test
+  @Ignore
```
Comment (anthonydahanne) [inline management/testing/entity-tests/src/test/java/org/terracotta/management/entity/sample/HelloWorldTest.java]: everything ignored / commented ?

File: management/management-registry/src/main/java/org/terracotta/management/registry/action/AbstractActionManagementProvider.java [hunk 5]
```diff
@@ -67,7 +74,9 @@ public AbstractActionManagementProvider(Class<? extends T> managedType) {
         }
       }
     }
-    return descriptors;
+    List<CallDescriptor> list = new ArrayList<CallDescriptor>(descriptors);
```
Comment (anthonydahanne) [inline management/management-registry/src/main/java/org/terracotta/management/registry/action/AbstractActionManagementProvider.java]: directly use a new TreeSet(CALL_DESCRIPTOR_COMPARATOR) instead of instantiating a set + a list ?

File: management/management-registry/src/test/java/org/terracotta/management/registry/ManagementRegistryTest.java [hunk 6]
```diff
@@ -66,8 +61,8 @@ public ContextContainer getContextContainer() {
     mapper.addMixIn(CapabilityContext.class, CapabilityContextMixin.class);
 
     String actual = mapper.writeValueAsString(registry.getCapabilities());
-    System.out.println(expectedJson);
-    System.out.println(actual);
+    //System.out.println(expectedJson);
+    //System.out.println(actual);
```
Comment (anthonydahanne) [inline management/management-registry/src/test/java/org/terracotta/management/registry/ManagementRegistryTest.java]: remove comments ?

File: management/monitoring-service-api/src/main/java/org/terracotta/management/service/monitoring/registry/provider/AbstractStatisticsManagementProvider.java [hunk 7]
```diff
@@ -66,8 +75,19 @@ public Capability getCapability() {
   }
 
   @Override
+  public final Collection<? extends Descriptor> getDescriptors() {
+    Collection<StatisticDescriptor> capabilities = new HashSet<>();
+    for (ExposedObject o : getExposedObjects()) {
+      capabilities.addAll(((AbstractExposedStatistics<?>) o).getDescriptors());
+    }
+    List<StatisticDescriptor> list = new ArrayList<>(capabilities);
+    Collections.sort(list, STATISTIC_DESCRIPTOR_COMPARATOR);
```
Comment (anthonydahanne) [inline management/monitoring-service-api/src/main/java/org/terracotta/management/service/monitoring/registry/provider/AbstractStatisticsManagementProvider.java]: why not using new TreeSet(STATISTIC_DESCRIPTOR_COMPARATOR) ?

File: management/testing/sample-entity/pom.xml [hunk 8]
```diff
@@ -97,9 +101,15 @@
...
+      <groupId>ch.qos.logback</groupId>
+      <artifactId>logback-classic</artifactId>
+      <version>${logback.version}</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
       <groupId>com.fasterxml.jackson.core</groupId>
       <artifactId>jackson-databind</artifactId>
-      <version>2.8.0</version>
+      <version>2.6.3</version>
```
Comment (anthonydahanne) [inline management/testing/sample-entity/pom.xml]: downgrade ? why ?

File: management/testing/pom.xml (hunk 1 above)
Comment (mathieucarbou) [inline management/testing/pom.xml]: I'll remove - no need to have the parent pom. I will do like distribution. Also, in the tc platform project we never have some artifact called -parent event if we have some parents (see management entity and tms entity for example). So -parent would be inconsistent with what we have.

File: management/distribution/management-server/pom.xml (hunk 2 above)
Comment (mathieucarbou) [inline management/distribution/management-server/pom.xml]: no its the right one ;-) this avoids the common libs to be shaded

File: management/management-registry/src/main/java/org/terracotta/management/registry/collect/StatisticsRegistryMetadata.java (hunk 3 above)
Comment (mathieucarbou) [inline management/management-registry/src/main/java/org/terracotta/management/registry/collect/StatisticsRegistryMetadata.java]: in the sample app :-) sizes are ints

File: management/testing/entity-tests/src/test/java/org/terracotta/management/entity/sample/HelloWorldTest.java (hunk 4 above)
Comment (mathieucarbou) [inline management/testing/entity-tests/src/test/java/org/terracotta/management/entity/sample/HelloWorldTest.java]: All what is inside entity-tests will be removed: these are old tests.

File: management/management-registry/src/main/java/org/terracotta/management/registry/action/AbstractActionManagementProvider.java (hunk 5 above)
Comment (mathieucarbou) [inline management/management-registry/src/main/java/org/terracotta/management/registry/action/AbstractActionManagementProvider.java]: no its not the same actually. I tried. Objects in management have a equals/hashcode. 
TreeSet do not check equals hashcode but the comparstor.

What I need is to add some objects and be sure they are unique as per their eq/hash and then sort them as per one of their property.

using treeset would discard some objects if the property is the same but not the eq/hash

File: management/monitoring-service-api/src/main/java/org/terracotta/management/service/monitoring/registry/provider/AbstractStatisticsManagementProvider.java (hunk 7 above)
Comment (mathieucarbou) [inline management/monitoring-service-api/src/main/java/org/terracotta/management/service/monitoring/registry/provider/AbstractStatisticsManagementProvider.java]: same answer as above ;-)

File: management/testing/sample-entity/pom.xml (hunk 8 above)
Comment (mathieucarbou) [inline management/testing/sample-entity/pom.xml]: oh its a mistake!

Comment (mathieucarbou) [inline management/testing/sample-entity/pom.xml]: oh i know! its a copy paste : every where in the project this is the same version: 2.6.3

Comment (mathieucarbou) [inline management/testing/sample-entity/pom.xml]: ok I remember! 2.6.3 supports 1.6 not 2.8

**Comment by mathieucarbou** [comment]
@anthonydahanne FYI: rebased on master, good to go after the release for #198 !

File: management/testing/sample-entity/pom.xml (hunk 8 above)
Comment (anthonydahanne) [inline management/testing/sample-entity/pom.xml]: well, tc platform is java 1.8 so .... why should we care about a 3rd party dep compatible with 1.6 ?

Comment (mathieucarbou) [inline management/testing/sample-entity/pom.xml]: Tc platform isb1.6 also. It is in ehcache client.

Comment (anthonydahanne) [inline management/testing/sample-entity/pom.xml]: well, out of the 4 modules using jackson-databind (management-registry, monitoring-service, cluster-topology, and sample-entity) only 1 can be deployed in ehcache3...
anyway, I guess that's fine, as long as we don't seem to need 2.8 features...

Comment (mathieucarbou) [inline management/testing/sample-entity/pom.xml]: and its only for 4 lines total in some tests ;-) so quite un-relevant for runtime ;-)
